# Amooora text data cleaning

## Import libraries and load csv

In [1]:
import pandas as pd
import numpy as np
import string
from nltk.tokenize import word_tokenize
from nltk.stem.wordnet import WordNetLemmatizer
from sklearn.feature_extraction.text import TfidfVectorizer
import nltk
from nltk.corpus import stopwords
# nltk.download('punkt_tab')

In [2]:
!tree ..

..
├── notebooks
│   ├── dedemenezes_exploratory.ipynb
│   └── dedemenezes_vectorizing_and_lda.ipynb
├── raw_data
│   ├── combined_text_preprocessed.csv
│   ├── okcupid_profiles.csv
│   ├── okcupid_profiles_encoded.csv
│   ├── parsed_data_public.csv
│   ├── question_data.csv
│   └── user_data_public.csv
├── README.md
└── requirements.txt

2 directories, 10 files


In [3]:
df = pd.read_csv('../raw_data/okcupid_profiles.csv')

In [4]:
df.shape

(59946, 31)

In [5]:
df.head(2)

age  status sex orientation       body_type               diet    drinks  \
0   22  single   m    straight  a little extra  strictly anything  socially   
1   35  single   m    straight         average       mostly other     often   

       drugs                      education     ethnicity  ...  \
0      never  working on college/university  asian, white  ...   
1  sometimes          working on space camp         white  ...   

                                              essay0  \
0  about me:  i would love to think that i was so...   
1  i am a chef: this is what that means. 1. i am ...   

                                              essay1  \
0  currently working as an international agent fo...   
1  dedicating everyday to being an unbelievable b...   

                                              essay2  \
0  making people laugh. ranting about a good salt...   
1  being silly. having ridiculous amonts of fun w...   

                                              essay3  \
0  the way i look. i am a six foot half asian, ha...   
1                                                NaN   

                                              essay4  \
0  books: absurdistan, the republic, of mice and ...   
1  i am die hard christopher moore fan. i don't r...   

                                              essay5  \
0                  food. water. cell phone. shelter.   
1  delicious porkness in all of its glories. my b...   

                        essay6  \
0  duality and humorous things   
1                          NaN   

                                              essay7  \
0  trying to find someone to hang out with. i am ...   
1                                                NaN   

                                              essay8  \
0  i am new to california and looking for someone...   
1  i am very open and will share just about anyth...   

                                              essay9  
0  you want to be swept off your feet! you are ti...  
1                                                NaN  

[2 rows x 31 columns]

## Access Text Dataset

In [6]:
df.iloc[:, :21]

age     status sex orientation       body_type               diet  \
0       22     single   m    straight  a little extra  strictly anything   
1       35     single   m    straight         average       mostly other   
2       38  available   m    straight            thin           anything   
3       23     single   m    straight            thin         vegetarian   
4       29     single   m    straight        athletic                NaN   
...    ...        ...  ..         ...             ...                ...   
59941   59     single   f    straight             NaN                NaN   
59942   24     single   m    straight             fit    mostly anything   
59943   42     single   m    straight         average    mostly anything   
59944   27     single   m    straight        athletic    mostly anything   
59945   39     single   m         gay         average                NaN   

           drinks      drugs                          education  \
0        socially      never      working on college/university   
1           often  sometimes              working on space camp   
2        socially        NaN     graduated from masters program   
3        socially        NaN      working on college/university   
4        socially      never  graduated from college/university   
...           ...        ...                                ...   
59941    socially      never  graduated from college/university   
59942       often  sometimes      working on college/university   
59943  not at all      never     graduated from masters program   
59944    socially      often      working on college/university   
59945    socially        NaN     graduated from masters program   

                 ethnicity  ...  income                           job  \
0             asian, white  ...      -1                transportation   
1                    white  ...   80000          hospitality / travel   
2                      NaN  ...      -1                           NaN   
3                    white  ...   20000                       student   
4      asian, black, other  ...      -1   artistic / musical / writer   
...                    ...  ...     ...                           ...   
59941                  NaN  ...      -1   sales / marketing / biz dev   
59942         white, other  ...      -1         entertainment / media   
59943                asian  ...  100000  construction / craftsmanship   
59944         asian, black  ...      -1             medicine / health   
59945                white  ...      -1             medicine / health   

            last_online                         location  \
0      2012-06-28-20-30  south san francisco, california   
1      2012-06-29-21-41              oakland, california   
2      2012-06-27-09-10        san francisco, california   
3      2012-06-28-14-22             berkeley, california   
4      2012-06-27-21-26        san francisco, california   
...                 ...                              ...   
59941  2012-06-12-21-47              oakland, california   
59942  2012-06-29-11-01        san francisco, california   
59943  2012-06-27-23-37  south san francisco, california   
59944  2012-06-23-13-01        san francisco, california   
59945  2012-06-29-00-42        san francisco, california   

                                    offspring                       pets  \
0      doesn't have kids, but might want them  likes dogs and likes cats   
1      doesn't have kids, but might want them  likes dogs and likes cats   
2                                         NaN                   has cats   
3                           doesn't want kids                 likes cats   
4                                         NaN  likes dogs and likes cats   
...                                       ...                        ...   
59941                                has kids                   has dogs   
59942                       doesn't have kids  likes dogs and likes cats   
59943  

In [7]:
text_df = df.iloc[:, 21:]

In [8]:
text_df.head()

essay0  \
0  about me:  i would love to think that i was so...   
1  i am a chef: this is what that means. 1. i am ...   
2  i'm not ashamed of much, but writing public te...   
3          i work in a library and go to school. . .   
4  hey how's it going? currently vague on the pro...   

                                              essay1  \
0  currently working as an international agent fo...   
1  dedicating everyday to being an unbelievable b...   
2  i make nerdy software for musicians, artists, ...   
3          reading things written by old dead people   
4                         work work work work + play   

                                              essay2  \
0  making people laugh. ranting about a good salt...   
1  being silly. having ridiculous amonts of fun w...   
2  improvising in different contexts. alternating...   
3  playing synthesizers and organizing books acco...   
4  creating imagery to look at: http://bagsbrown....   

                                              essay3  \
0  the way i look. i am a six foot half asian, ha...   
1                                                NaN   
2  my large jaw and large glasses are the physica...   
3                  socially awkward but i do my best   
4            i smile a lot and my inquisitive nature   

                                              essay4  \
0  books: absurdistan, the republic, of mice and ...   
1  i am die hard christopher moore fan. i don't r...   
2  okay this is where the cultural matrix gets so...   
3  bataille, celine, beckett. . . lynch, jarmusch...   
4  music: bands, rappers, musicians at the moment...   

                                              essay5  \
0                  food. water. cell phone. shelter.   
1  delicious porkness in all of its glories. my b...   
2  movement conversation creation contemplation t...   
3                                                NaN   
4                                                NaN   

                        essay6  \
0  duality and humorous things   
1                          NaN   
2                          NaN   
3   cats and german philosophy   
4                          NaN   

                                              essay7  \
0  trying to find someone to hang out with. i am ...   
1                                                NaN   
2  viewing. listening. dancing. talking. drinking...   
3                                                NaN   
4                                                NaN   

                                              essay8  \
0  i am new to california and looking for someone...   
1  i am very open and will share just about anyth...   
2  when i was five years old, i was known as "the...   
3                                                NaN   
4                                                NaN   

                                              essay9  
0  you want to be swept off your feet! you are ti...  
1                                                NaN  
2  you are bright, open, intense, silly, ironic, ...  
3                              you feel so inclined.  
4                                                NaN

## Checking dataset

In [9]:
text_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59946 entries, 0 to 59945
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   essay0  54458 non-null  object
 1   essay1  52374 non-null  object
 2   essay2  50308 non-null  object
 3   essay3  48470 non-null  object
 4   essay4  49409 non-null  object
 5   essay5  49096 non-null  object
 6   essay6  46175 non-null  object
 7   essay7  47495 non-null  object
 8   essay8  40721 non-null  object
 9   essay9  47343 non-null  object
dtypes: object(10)
memory usage: 4.6+ MB


### Duplicates

In [10]:
text_df.duplicated().sum()

2134

In [11]:
text_df[text_df.duplicated()]

essay0 essay1 essay2 essay3 essay4 essay5 essay6 essay7 essay8 essay9
51       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
64       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
80       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
84       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
175      NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...
59757    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
59791    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
59857    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
59881    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN
59930    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN

[2134 rows x 10 columns]

We have `2134`duplicated observations all made of empty answers to all questions

In [12]:
# no_dup_df = text_df.drop_duplicates()
# no_dup_df.shape

### Empty values

In [13]:
text_df.isnull().sum().sort_values(ascending=False) / len(text_df) * 100

essay8    32.070530
essay6    22.972342
essay9    21.023922
essay7    20.770360
essay3    19.143896
essay5    18.099623
essay4    17.577486
essay2    16.077803
essay1    12.631368
essay0     9.154906
dtype: float64

In [14]:
text_df.dropna().shape

(29866, 10)

In [15]:
text_df.shape

(59946, 10)

Drop is not an option!

## % of answered questions

In [16]:
text_df.columns

Index(['essay0', 'essay1', 'essay2', 'essay3', 'essay4', 'essay5', 'essay6',
       'essay7', 'essay8', 'essay9'],
      dtype='object')

In [17]:
text_df['not_answered_percent'] = text_df.apply(lambda x: x.isna().sum() / len(text_df.columns) * 100, axis=1)

In [18]:
text_df

essay0  \
0      about me:  i would love to think that i was so...   
1      i am a chef: this is what that means. 1. i am ...   
2      i'm not ashamed of much, but writing public te...   
3              i work in a library and go to school. . .   
4      hey how's it going? currently vague on the pro...   
...                                                  ...   
59941  vibrant, expressive, caring optimist. i love b...   
59942  i'm nick. i never know what to write about mys...   
59943  hello! i enjoy traveling, watching movies, and...   
59944  "all i have in this world are my balls and my ...   
59945  is it odd that having a little "enemy" status ...   

                                                  essay1  \
0      currently working as an international agent fo...   
1      dedicating everyday to being an unbelievable b...   
2      i make nerdy software for musicians, artists, ...   
3              reading things written by old dead people   
4                             work work work work + play   
...                                                  ...   
59941  the happiest times have been when life came to...   
59942  currently finishing school for film production...   
59943  i'm a civil engineer, who enjoys helping the c...   
59944  following my dreams... "you got a dream... you...   
59945  i work with elderly people (psychotherapy and ...   

                                                  essay2  \
0      making people laugh. ranting about a good salt...   
1      being silly. having ridiculous amonts of fun w...   
2      improvising in different contexts. alternating...   
3      playing synthesizers and organizing books acco...   
4      creating imagery to look at: http://bagsbrown....   
...                                                  ...   
59941  i make an outstanding osso bucco. i am also ve...   
59942  filmmaking, photography, graphic design, web d...   
59943  - looking at things objectively - getting thin...   
59944                                          listening   
59945  i'm a great bullshitter. i don't know what it ...   

                                                  essay3  \
0      the way i look. i am a six foot half asian, ha...   
1                                                    NaN   
2      my large jaw and large glasses are the physica...   
3                      socially awkward but i do my best   
4                i smile a lot and my inquisitive nature   
...                                                  ...   
59941  i am told that people notice my smile, eyes an...   
59942                                dude, i don't know.   
59943  i'm quiet until i get used to the environment ...   
59944  it used to be the hair until i mowed it off bu...   
59945  either that i am funny/sarcastic, or that i am...   

                                                  essay4  \
0      books: absurdistan, the republic, of mice and ...   
1      i am die hard christopher moore fan. i don't r...   
2      okay this is where the cultural matrix gets so...   
3      bataille, celine, beckett. . . lynch, jarmusch...   
4      music: bands, rappers, musicians at the moment...   
...                                                  ...   
59941  i am an avid movie watcher and follow the broa...   
59942  movies: hook (the greatest adventure ever!), g...   
59943  last book: "game change". movies: bourne serie...   
59944  where to begin musically: right now i listen t...   
59945  i just read the help by kathryn stockett, sooo...   

                                                  essay5  \
0                      food. water. cell phone. shelter.   
1      delicious porkness in all of its glories. my b...   
2      movement conversation creation contemplation t...   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
59941         my family, 

## Replace NaN with empty string

In [19]:
no_nan_df = text_df.replace(np.nan, '')

In [20]:
no_nan_df.isna()

essay0  essay1  essay2  essay3  essay4  essay5  essay6  essay7  essay8  \
0       False   False   False   False   False   False   False   False   False   
1       False   False   False   False   False   False   False   False   False   
2       False   False   False   False   False   False   False   False   False   
3       False   False   False   False   False   False   False   False   False   
4       False   False   False   False   False   False   False   False   False   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
59941   False   False   False   False   False   False   False   False   False   
59942   False   False   False   False   False   False   False   False   False   
59943   False   False   False   False   False   False   False   False   False   
59944   False   False   False   False   False   False   False   False   False   
59945   False   False   False   False   False   False   False   False   False   

       essay9  not_answered_percent  
0       False                 False  
1       False                 False  
2       False                 False  
3       False                 False  
4       False                 False  
...       ...                   ...  
59941   False                 False  
59942   False                 False  
59943   False                 False  
59944   False                 False  
59945   False                 False  

[59946 rows x 11 columns]

## Combining into one big text column

In [21]:
no_nan_df['combined'] = no_nan_df.apply(lambda row: ' '.join(row.values.astype(str)), axis=1)

In [22]:
no_nan_df.combined

0        about me:  i would love to think that i was so...
1        i am a chef: this is what that means. 1. i am ...
2        i'm not ashamed of much, but writing public te...
3        i work in a library and go to school. . . read...
4        hey how's it going? currently vague on the pro...
                               ...                        
59941    vibrant, expressive, caring optimist. i love b...
59942    i'm nick. i never know what to write about mys...
59943    hello! i enjoy traveling, watching movies, and...
59944    "all i have in this world are my balls and my ...
59945    is it odd that having a little "enemy" status ...
Name: combined, Length: 59946, dtype: object

## Cleaning text

In [23]:
def basic_cleaning(sentence):
    # Basic cleaning
    sentence = sentence.strip() # remove whitespaces
    sentence = sentence.lower() # lowercase
    sentence = ''.join(char for char in sentence if not char.isdigit()) # remove numbers
    
    # Advanced cleaning
    for punctuation in string.punctuation:
        sentence = sentence.replace(punctuation, '')
        
    return sentence

In [24]:
%%time
cleaned_df = pd.DataFrame()

# Criar Novas colunas essayX_clean
essays = no_nan_df.columns[:10]
for essay in essays:
    cleaned_df[f"{essay}_clean"] = no_nan_df[essay].apply(basic_cleaning)

cleaned_df['combined_clean'] = no_nan_df.combined.apply(basic_cleaning)

cleaned_df

CPU times: user 17.6 s, sys: 217 ms, total: 17.8 s
Wall time: 17.3 s


essay0_clean  \
0      about me  i would love to think that i was som...   
1      i am a chef this is what that means  i am a wo...   
2      im not ashamed of much but writing public text...   
3                 i work in a library and go to school     
4      hey hows it going currently vague on the profi...   
...                                                  ...   
59941  vibrant expressive caring optimist i love bein...   
59942  im nick i never know what to write about mysel...   
59943  hello i enjoy traveling watching movies and ha...   
59944  all i have in this world are my balls and my i...   
59945  is it odd that having a little enemy status wi...   

                                            essay1_clean  \
0      currently working as an international agent fo...   
1      dedicating everyday to being an unbelievable b...   
2      i make nerdy software for musicians artists an...   
3              reading things written by old dead people   
4                              work work work work  play   
...                                                  ...   
59941  the happiest times have been when life came to...   
59942  currently finishing school for film production...   
59943  im a civil engineer who enjoys helping the cit...   
59944  following my dreams you got a dream you gotta ...   
59945  i work with elderly people psychotherapy and c...   

                                            essay2_clean  \
0      making people laugh ranting about a good salti...   
1      being silly having ridiculous amonts of fun wh...   
2      improvising in different contexts alternating ...   
3      playing synthesizers and organizing books acco...   
4      creating imagery to look at httpbagsbrownblogs...   
...                                                  ...   
59941  i make an outstanding osso bucco i am also ver...   
59942  filmmaking photography graphic design web desi...   
59943   looking at things objectively  getting things...   
59944                                          listening   
59945  im a great bullshitter i dont know what it is ...   

                                            essay3_clean  \
0      the way i look i am a six foot half asian half...   
1                                                          
2      my large jaw and large glasses are the physica...   
3                      socially awkward but i do my best   
4                i smile a lot and my inquisitive nature   
...                                                  ...   
59941  i am told that people notice my smile eyes and...   
59942                                   dude i dont know   
59943  im quiet until i get used to the environment b...   
59944  it used to be the hair until i mowed it off bu...   
59945  either that i am funnysarcastic or that i am t...   

                                            essay4_clean  \
0      books absurdistan the republic of mice and men...   
1      i am die hard christopher moore fan i dont rea...   
2      okay this is where the cultural matrix gets so...   
3      bataille celine beckett   lynch jarmusch rw fa...   
4      music bands rappers musicians at the moment th...   
...                                                  ...   
59941  i am an avid movie watcher and follow the broa...   
59942  movies hook the greatest adventure ever gladia...   
59943  last book game change movies bourne series act...   
59944  where to begin musically right now i listen to...   
59945  i just read the help by kathryn stockett soooo...   

                                            essay5_clean  \
0                          food water cell phone shelter   
1      delicious porkness in all of its glories my bi...   
2      movement conversation creation contemplation t...   
3                                                          
4                                                          
...                                                  ...   
59941             m

## Preprocessing

### Tokenization

Let's take one row and play with **Tokenization**, **stopwords** and **lemmatize** 

In [25]:
text = no_nan_df.loc[0, 'essay0']

word_tokens = word_tokenize(text)

pd.DataFrame({'essay0_words': word_tokens})

essay0_words
0          about
1             me
2              :
3              i
4          would
..           ...
264            a
265         good
266    attention
267         span
268            .

[269 rows x 1 columns]

### Remove stopwords

In [26]:
word_tokens

stop_words = set(stopwords.words('english'))

pd.DataFrame()

stopwords_removed = [w for w in word_tokens if w in stop_words]

tokens_cleaned = [w for w in word_tokens if not w in stop_words]
tokens_cleaned

[':',
 'would',
 'love',
 'think',
 'kind',
 'intellectual',
 ':',
 'either',
 'dumbest',
 'smart',
 'guy',
 ',',
 'smartest',
 'dumb',
 'guy',
 '.',
 'ca',
 "n't",
 'say',
 'tell',
 'difference',
 '.',
 'love',
 'talk',
 'ideas',
 'concepts',
 '.',
 'forge',
 'odd',
 'metaphors',
 'instead',
 'reciting',
 'cliches',
 '.',
 'like',
 'simularities',
 'friend',
 'mine',
 "'s",
 'house',
 'underwater',
 'salt',
 'mine',
 '.',
 'favorite',
 'word',
 'salt',
 'way',
 '(',
 'weird',
 'choice',
 'know',
 ')',
 '.',
 'things',
 'life',
 'better',
 'metaphors',
 '.',
 'seek',
 'make',
 'little',
 'better',
 'everyday',
 ',',
 'productively',
 'lazy',
 'way',
 '.',
 'got',
 'tired',
 'tying',
 'shoes',
 '.',
 'considered',
 'hiring',
 'five',
 'year',
 'old',
 ',',
 'would',
 'probably',
 'tie',
 'shoes',
 '...',
 'decided',
 'wear',
 'leather',
 'shoes',
 'dress',
 'shoes',
 '.',
 ':',
 'love',
 'really',
 'serious',
 ',',
 'really',
 'deep',
 'conversations',
 'really',
 'silly',
 'stuff',
 '.

### Lemmatizing

In [27]:
# Lemmatizing the verbs
verb_lemmatized = [
    WordNetLemmatizer().lemmatize(word, pos = "v") # v --> verbs
    for word in tokens_cleaned
]

# 2 - Lemmatizing the nouns
noun_lemmatized = [
    WordNetLemmatizer().lemmatize(word, pos = "n") # n --> nouns
    for word in verb_lemmatized
]

pd.DataFrame(dict(token=tokens_cleaned, verbs=verb_lemmatized, noun=noun_lemmatized))

token      verbs       noun
0            :          :          :
1        would      would      would
2         love       love       love
3        think      think      think
4         kind       kind       kind
..         ...        ...        ...
150     lastly     lastly     lastly
151       good       good       good
152  attention  attention  attention
153       span       span       span
154          .          .          .

[155 rows x 3 columns]

### All together

In [28]:
def basic_preprocessing(sentence):
    tokenized_sentence = word_tokenize(sentence) ## tokenize
    stop_words = set(stopwords.words('english')) ## define stopwords

    tokenized_sentence_cleaned = [ ## remove stopwords
        w for w in tokenized_sentence if not w in stop_words
    ]

    lemmatized = [
        WordNetLemmatizer().lemmatize(word, pos = "v")
        for word in tokenized_sentence_cleaned
    ]

    # 2 - Lemmatizing the nouns
    noun_lemmatized = [
        WordNetLemmatizer().lemmatize(word, pos = "n")
        for word in lemmatized
    ]
    
    cleaned_sentence = ' '.join(word for word in noun_lemmatized)
    return cleaned_sentence

text = cleaned_df.loc[0, 'essay0_clean']
print(f"Original text:\n{text}")
basic_preprocessing(text)

Original text:
about me  i would love to think that i was some some kind of intellectual either the dumbest smart guy or the smartest dumb guy cant say i can tell the difference i love to talk about ideas and concepts i forge odd metaphors instead of reciting cliches like the simularities between a friend of mines house and an underwater salt mine my favorite word is salt by the way weird choice i know to me most things in life are better as metaphors i seek to make myself a little better everyday in some productively lazy way got tired of tying my shoes considered hiring a five year old but would probably have to tie both of our shoes decided to only wear leather shoes dress shoes  about you  you love to have really serious really deep conversations about really silly stuff you have to be willing to snap me out of a light hearted rant with a kiss you dont have to be funny but you have to be able to make me laugh you should be able to bend spoons with your mind and telepathically make 

'would love think kind intellectual either dumbest smart guy smartest dumb guy cant say tell difference love talk idea concept forge odd metaphor instead recite cliche like simularities friend mine house underwater salt mine favorite word salt way weird choice know thing life better metaphor seek make little better everyday productively lazy way get tire tie shoe consider hire five year old would probably tie shoe decide wear leather shoe dress shoe love really serious really deep conversation really silly stuff will snap light hearted rant kiss dont funny able make laugh able bend spoon mind telepathically make smile still work love life cool let wind blow extra point read guess favorite video game hint give yet lastly good attention span'

In [29]:
# cleaned_df.essay0_clean.apply(basic_preprocessing)

In [30]:
# %%time
# clean_preprocessed_df = cleaned_df.apply(lambda row: row.apply(basic_preprocessing))

In [31]:
# %%time
# clean_preprocessed_df = cleaned_df.applymap(basic_preprocessing)

In [32]:
%%time
clean_preprocessed_df = cleaned_df['combined_clean'].apply(basic_preprocessing)

CPU times: user 2min 11s, sys: 1.3 s, total: 2min 13s
Wall time: 2min 13s


In [33]:
clean_preprocessed_df

0        would love think kind intellectual either dumb...
1        chef mean workaholic love cook regardless whet...
2        im ashamed much write public text online date ...
3        work library go school read thing write old de...
4        hey hows go currently vague profile know come ...
                               ...                        
59941    vibrant expressive care optimist love people t...
59942    im nick never know write im sure hand im south...
59943    hello enjoy travel watch movie hang friend rul...
59944    world ball integrity one take either away momm...
59945    odd little enemy status someone make seem inte...
Name: combined_clean, Length: 59946, dtype: object

In [34]:
new_names = dict(
#     essay0_clean="essay0_preprocessed",
#     essay1_clean="essay1_preprocessed",
#     essay2_clean="essay2_preprocessed",
#     essay3_clean="essay3_preprocessed",
#     essay4_clean="essay4_preprocessed",
#     essay5_clean="essay5_preprocessed",
#     essay6_clean="essay6_preprocessed",
#     essay7_clean="essay7_preprocessed",
#     essay8_clean="essay8_preprocessed",
#     essay9_clean="essay9_preprocessed",
    combined_clean="combined_preprocessed"
)
clean_preprocessed_df = clean_preprocessed_df.to_frame().rename(columns=new_names)

In [35]:
clean_preprocessed_df.combined_preprocessed.head(1)

0    would love think kind intellectual either dumb...
Name: combined_preprocessed, dtype: object

In [36]:
cleaned_df.combined_clean.head(1)

0    about me  i would love to think that i was som...
Name: combined_clean, dtype: object

In [37]:
no_nan_df.combined.head(1)

0    about me:  i would love to think that i was so...
Name: combined, dtype: object

## Saving into CSV

In [38]:
!tree ..

..
├── notebooks
│   └── dedemenezes_exploratory.ipynb
├── raw_data
│   ├── combined_text_preprocessed.csv
│   ├── okcupid_profiles.csv
│   ├── okcupid_profiles_encoded.csv
│   ├── parsed_data_public.csv
│   ├── question_data.csv
│   └── user_data_public.csv
├── README.md
└── requirements.txt

2 directories, 9 files


In [39]:
clean_preprocessed_df.to_csv('../raw_data/combined_text_preprocessed.csv')

In [40]:
!tree ..

..
├── notebooks
│   └── dedemenezes_exploratory.ipynb
├── raw_data
│   ├── combined_text_preprocessed.csv
│   ├── okcupid_profiles.csv
│   ├── okcupid_profiles_encoded.csv
│   ├── parsed_data_public.csv
│   ├── question_data.csv
│   └── user_data_public.csv
├── README.md
└── requirements.txt

2 directories, 9 files


## Vectorizing

Using relative frequency rather than count is robust to document length

Takes into account the context of the whole corpus

In [38]:
clean_preprocessed_df.shape

(59946, 1)

In [39]:
clean_preprocessed_df.combined_preprocessed.values

array(['would love think kind intellectual either dumbest smart guy smartest dumb guy cant say tell difference love talk idea concept forge odd metaphor instead recite cliche like simularities friend mine house underwater salt mine favorite word salt way weird choice know thing life better metaphor seek make little better everyday productively lazy way get tire tie shoe consider hire five year old would probably tie shoe decide wear leather shoe dress shoe love really serious really deep conversation really silly stuff will snap light hearted rant kiss dont funny able make laugh able bend spoon mind telepathically make smile still work love life cool let wind blow extra point read guess favorite video game hint give yet lastly good attention span currently work international agent freight forward company import export domestic know work online class try better free time perhaps hour worth good book video game lazy sunday make people laugh rant good salt find simplicity complexity compl

In [40]:
%%time
# Instantiating the TfidfVectorizer
tf_idf_vectorizer = TfidfVectorizer(min_df=0.1, ngram_range=(1,2))

# Training it on the texts
combined_weighted_words = pd.DataFrame(tf_idf_vectorizer.fit_transform(clean_preprocessed_df.combined_preprocessed.values).toarray(),
                 columns = tf_idf_vectorizer.get_feature_names_out())

combined_weighted_words

CPU times: user 19.2 s, sys: 1.04 s, total: 20.3 s
Wall time: 19.9 s


actually  adventure    almost     along      also    always     amaze  \
0           0.0   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
1           0.0   0.000000  0.000000  0.000000  0.000000  0.117395  0.000000   
2           0.0   0.000000  0.000000  0.000000  0.081198  0.000000  0.000000   
3           0.0   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
4           0.0   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
...         ...        ...       ...       ...       ...       ...       ...   
59941       0.0   0.000000  0.000000  0.142332  0.078259  0.000000  0.000000   
59942       0.0   0.084036  0.193807  0.000000  0.056493  0.066949  0.000000   
59943       0.0   0.000000  0.103404  0.000000  0.060283  0.000000  0.000000   
59944       0.0   0.000000  0.000000  0.000000  0.000000  0.079786  0.000000   
59945       0.0   0.000000  0.000000  0.000000  0.046083  0.054612  0.082074   

       american  anything      area  ...     woman      word      work  \
0      0.000000  0.125752  0.000000  ...  0.000000  0.103616  0.140483   
1      0.000000  0.107094  0.000000  ...  0.000000  0.000000  0.079760   
2      0.000000  0.043892  0.000000  ...  0.000000  0.000000  0.032689   
3      0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.187798   
4      0.212428  0.000000  0.000000  ...  0.000000  0.000000  0.385680   
...         ...       ...       ...  ...       ...       ...       ...   
59941  0.000000  0.000000  0.000000  ...  0.000000  0.139426  0.126022   
59942  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.045486   
59943  0.000000  0.065172  0.181377  ...  0.000000  0.000000  0.048538   
59944  0.000000  0.218355  0.000000  ...  0.112628  0.000000  0.000000   
59945  0.000000  0.049821  0.000000  ...  0.000000  0.000000  0.074209   

          world     would     write      year       yes       yet  youre  
0      0.000000  0.133960  0.000000  0.064378  0.000000  0.106134    0.0  
1      0.000000  0.000000  0.000000  0.000000  0.000000  0.000000    0.0  
2      0.049425  0.000000  0.057229  0.089881  0.000000  0.000000    0.0  
3      0.000000  0.000000  0.328780  0.000000  0.000000  0.000000    0.0  
4      0.000000  0.000000  0.000000  0.000000  0.000000  0.000000    0.0  
...         ...       ...       ...       ...       ...       ...    ...  
59941  0.285816  0.000000  0.110315  0.000000  0.000000  0.000000    0.0  
59942  0.000000  0.000000  0.159267  0.062534  0.102665  0.103094    0.0  
59943  0.000000  0.069426  0.000000  0.133459  0.000000  0.110010    0.0  
59944  0.081961  0.155072  0.000000  0.000000  0.000000  0.000000    0.0  
59945  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000    0.0  

[59946 rows x 282 columns]

Could run a cross validation but we do not have the y

### Vectorizing only bigrams

In [44]:
%%time
# Instantiating the TfidfVectorizer
bigram_tf_idf_vectorizer = TfidfVectorizer(min_df=0.1, ngram_range=(2,2))

# Training it on the texts
bigram_combined_weighted_words = pd.DataFrame(bigram_tf_idf_vectorizer.fit_transform(clean_preprocessed_df.combined_preprocessed.values).toarray(),
                 columns = bigram_tf_idf_vectorizer.get_feature_names_out())

bigram_combined_weighted_words

CPU times: user 17 s, sys: 1.54 s, total: 18.5 s
Wall time: 19.2 s


bay area  family friend  friend family  get know  im pretty  \
0      0.000000        0.00000       0.000000       0.0        0.0   
1      0.000000        0.00000       0.000000       0.0        0.0   
2      0.000000        0.00000       0.000000       0.0        0.0   
3      0.000000        0.00000       0.000000       0.0        0.0   
4      0.000000        0.00000       0.000000       0.0        1.0   
...         ...            ...            ...       ...        ...   
59941  0.000000        0.00000       0.000000       0.0        0.0   
59942  0.000000        0.00000       0.000000       0.0        0.0   
59943  0.744222        0.00000       0.391524       0.0        0.0   
59944  0.000000        1.00000       0.000000       0.0        0.0   
59945  0.000000        0.68588       0.000000       0.0        0.0   

       make people  new thing  san francisco  sense humor  spend time  \
0              1.0        0.0       0.000000     0.000000    0.000000   
1              0.0        1.0       0.000000     0.000000    0.000000   
2              0.0        0.0       0.653205     0.000000    0.757181   
3              0.0        0.0       0.000000     0.000000    0.000000   
4              0.0        0.0       0.000000     0.000000    0.000000   
...            ...        ...            ...          ...         ...   
59941          0.0        0.0       0.000000     0.000000    0.000000   
59942          0.0        0.0       1.000000     0.000000    0.000000   
59943          0.0        0.0       0.358501     0.000000    0.000000   
59944          0.0        0.0       0.000000     0.000000    0.000000   
59945          0.0        0.0       0.000000     0.727715    0.000000   

       watch movie  
0         0.000000  
1         0.000000  
2         0.000000  
3         0.000000  
4         0.000000  
...            ...  
59941     0.000000  
59942     0.000000  
59943     0.405362  
59944     0.000000  
59945     0.000000  

[59946 rows x 11 columns]

## Latent Dirichlet Allocation

In [41]:
from sklearn.decomposition import LatentDirichletAllocation

### 💻 Finding the topics (2)

In [42]:
%%time

# Instantiate the LDA
n_components = 2
lda_model = LatentDirichletAllocation(n_components=n_components, max_iter = 100)

# Fit the LDA on the vectorized documents
lda_model.fit(combined_weighted_words)

CPU times: user 11min 13s, sys: 96.2 ms, total: 11min 13s
Wall time: 11min 12s


LatentDirichletAllocation(max_iter=100, n_components=2)

In [43]:
combined_weighted_words

actually  adventure    almost     along      also    always     amaze  \
0           0.0   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
1           0.0   0.000000  0.000000  0.000000  0.000000  0.117395  0.000000   
2           0.0   0.000000  0.000000  0.000000  0.081198  0.000000  0.000000   
3           0.0   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
4           0.0   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
...         ...        ...       ...       ...       ...       ...       ...   
59941       0.0   0.000000  0.000000  0.142332  0.078259  0.000000  0.000000   
59942       0.0   0.084036  0.193807  0.000000  0.056493  0.066949  0.000000   
59943       0.0   0.000000  0.103404  0.000000  0.060283  0.000000  0.000000   
59944       0.0   0.000000  0.000000  0.000000  0.000000  0.079786  0.000000   
59945       0.0   0.000000  0.000000  0.000000  0.046083  0.054612  0.082074   

       american  anything      area  ...     woman      word      work  \
0      0.000000  0.125752  0.000000  ...  0.000000  0.103616  0.140483   
1      0.000000  0.107094  0.000000  ...  0.000000  0.000000  0.079760   
2      0.000000  0.043892  0.000000  ...  0.000000  0.000000  0.032689   
3      0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.187798   
4      0.212428  0.000000  0.000000  ...  0.000000  0.000000  0.385680   
...         ...       ...       ...  ...       ...       ...       ...   
59941  0.000000  0.000000  0.000000  ...  0.000000  0.139426  0.126022   
59942  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.045486   
59943  0.000000  0.065172  0.181377  ...  0.000000  0.000000  0.048538   
59944  0.000000  0.218355  0.000000  ...  0.112628  0.000000  0.000000   
59945  0.000000  0.049821  0.000000  ...  0.000000  0.000000  0.074209   

          world     would     write      year       yes       yet  youre  
0      0.000000  0.133960  0.000000  0.064378  0.000000  0.106134    0.0  
1      0.000000  0.000000  0.000000  0.000000  0.000000  0.000000    0.0  
2      0.049425  0.000000  0.057229  0.089881  0.000000  0.000000    0.0  
3      0.000000  0.000000  0.328780  0.000000  0.000000  0.000000    0.0  
4      0.000000  0.000000  0.000000  0.000000  0.000000  0.000000    0.0  
...         ...       ...       ...       ...       ...       ...    ...  
59941  0.285816  0.000000  0.110315  0.000000  0.000000  0.000000    0.0  
59942  0.000000  0.000000  0.159267  0.062534  0.102665  0.103094    0.0  
59943  0.000000  0.069426  0.000000  0.133459  0.000000  0.110010    0.0  
59944  0.081961  0.155072  0.000000  0.000000  0.000000  0.000000    0.0  
59945  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000    0.0  

[59946 rows x 282 columns]

#### Document Mixture (of Topics)

In [44]:
%%time
document_topic_mixture = lda_model.transform(combined_weighted_words)

document_topic_mixture

pd.DataFrame({
    'topic_0': document_topic_mixture[:, 0],
    'topic_1': document_topic_mixture[:, 1],
    'original_text': clean_preprocessed_df.combined_preprocessed
})

CPU times: user 4.45 s, sys: 19.1 ms, total: 4.47 s
Wall time: 4.64 s


topic_0   topic_1                                      original_text
0      0.274183  0.725817  would love think kind intellectual either dumb...
1      0.452609  0.547391  chef mean workaholic love cook regardless whet...
2      0.279403  0.720597  im ashamed much write public text online date ...
3      0.308972  0.691028  work library go school read thing write old de...
4      0.446582  0.553418  hey hows go currently vague profile know come ...
...         ...       ...                                                ...
59941  0.579179  0.420821  vibrant expressive care optimist love people t...
59942  0.211826  0.788174  im nick never know write im sure hand im south...
59943  0.704966  0.295034  hello enjoy travel watch movie hang friend rul...
59944  0.309418  0.690582  world ball integrity one take either away momm...
59945  0.231129  0.768871  odd little enemy status someone make seem inte...

[59946 rows x 3 columns]

#### Topic Mixture (of Words)

In [45]:
topic_word_mixture = pd.DataFrame(
    lda_model.components_,
    columns = tf_idf_vectorizer.get_feature_names_out()
)

topic_word_mixture

actually    adventure      almost       along         also       always  \
0    0.997971  1185.783303  253.667895  201.087691   890.964821   802.123577   
1  844.710070    90.612647  542.994245  462.789140  1809.841124  1040.189805   

        amaze    american     anything         area  ...       woman  \
0  403.609497  304.684356   763.754527  1160.744266  ...  335.512743   
1  348.761825  461.369178  1522.259945     1.062497  ...  733.091187   

         word        work        world        would        write         year  \
0    0.823959  2629.75969   756.278986   573.463388     1.761021  1293.294040   
1  769.965426  1231.35007  1058.305126  1468.466932  1514.353265   869.843244   

          yes         yet        youre  
0    1.141245   19.238983   111.604404  
1  701.360085  673.360466  1550.491331  

[2 rows x 282 columns]

##### Most relevant words for each topic

In [46]:
topic_word_mixture.iloc[0].sort_values(ascending = False).head(10)

love      4152.329831
friend    3911.746220
enjoy     2757.260691
new       2756.515908
work      2629.759690
family    2496.098706
music     2490.219367
food      2354.990138
good      2270.869789
life      2170.291793
Name: 0, dtype: float64

In [47]:
topic_word_mixture.iloc[1].sort_values(ascending = False).head(10)

im        6409.862066
like      3926.870886
get       2976.164029
dont      2926.147280
thing     2763.980983
think     2704.523241
really    2583.148650
know      2542.088579
one       2299.827623
people    2283.839050
Name: 1, dtype: float64

In [48]:
lda_model.score(combined_weighted_words)

-2221820.6057441426

### Finding the topics (5)

In [49]:
%%time

n_components = 5
five_lda_model = LatentDirichletAllocation(n_components=n_components, max_iter = 100)


five_lda_model.fit(combined_weighted_words)

CPU times: user 19min 40s, sys: 120 ms, total: 19min 40s
Wall time: 19min 40s


LatentDirichletAllocation(max_iter=100, n_components=5)

#### Document Mixture of Topics

In [50]:
%%time
five_document_topic_mixture = five_lda_model.transform(combined_weighted_words)


CPU times: user 11.3 s, sys: 13 μs, total: 11.3 s
Wall time: 11.7 s


In [51]:
five_topics_df = pd.DataFrame({
    'topic_0': five_document_topic_mixture[:, 0],
    'topic_1': five_document_topic_mixture[:, 1],
    'topic_2': five_document_topic_mixture[:, 2],
    'topic_3': five_document_topic_mixture[:, 3],
    'topic_4': five_document_topic_mixture[:, 4],
})

#### Topic means

In [52]:
topic_means = five_topics_df.mean()
print(topic_means)

topic_0    0.168469
topic_1    0.179850
topic_2    0.183076
topic_3    0.168803
topic_4    0.299802
dtype: float64


#### Topic Mixture of Words

In [53]:
five_topic_word_mixture = pd.DataFrame(
    five_lda_model.components_,
    columns = tf_idf_vectorizer.get_feature_names_out()
)

five_topic_word_mixture

actually   adventure      almost       along         also      always  \
0   94.520191  273.212643  137.391514   94.556646   420.708358  310.799200   
1  122.241570  278.280351  139.550705   88.265074   436.247868  292.640167   
2   87.897301  338.008222  112.602688  119.382055   480.629126  322.468816   
3   54.627418  121.495697  104.774384  102.678998   322.452586  276.429682   
4  486.421561  265.399037  302.342848  258.994057  1040.768005  639.975517   

        amaze    american    anything        area  ...       woman  \
0  167.499175  101.815565  323.951043  588.308838  ...   61.750708   
1  120.785644  291.941543  476.212795   96.265194  ...  112.512290   
2  168.924757   96.050674  300.396385  160.535474  ...  419.328946   
3   48.207385   69.585522  386.490510   70.142001  ...  136.867676   
4  246.954362  206.660231  798.963739  246.555256  ...  338.144310   

         word         work       world       would       write        year  \
0   51.993945   761.019873  280.809503  252.420490  138.036143  615.471253   
1  144.300460   609.113291  338.493871  237.413026  331.522881  315.887277   
2  115.074526   634.187041  481.823304  449.725206  249.394721  338.345984   
3   32.096116   776.992129  130.754764  298.308385  102.313944  203.776126   
4  427.324337  1079.797426  582.702671  804.063212  694.846597  689.656645   

          yes         yet       youre  
0   77.918493   79.122208  203.289778  
1  119.704254   82.815055  302.311942  
2  105.204373  163.450999  195.977648  
3   43.126001   50.306214  163.073956  
4  356.548209  316.904974  797.442413  

[5 rows x 282 columns]

### Finding the topics (10) 

In [54]:
%%time

n_components = 10
ten_lda_model = LatentDirichletAllocation(n_components=n_components, max_iter = 100)


ten_lda_model.fit(combined_weighted_words)

CPU times: user 14min 55s, sys: 150 ms, total: 14min 55s
Wall time: 14min 55s


LatentDirichletAllocation(max_iter=100)

#### Document Mixture of Topics

In [55]:
%%time
ten_document_topic_mixture = ten_lda_model.transform(combined_weighted_words)

ten_topics = {}
for index in range(ten_document_topic_mixture.shape[1]):
    ten_topics[f"topic_{index}"] = ten_document_topic_mixture[:, index]
    
ten_topics

CPU times: user 8.23 s, sys: 0 ns, total: 8.23 s
Wall time: 7.24 s


{'topic_0': array([0.01144798, 0.01335621, 0.01031735, ..., 0.0117379 , 0.01183464,
        0.01073903]),
 'topic_1': array([0.01144787, 0.01335477, 0.01031793, ..., 0.01173906, 0.01183428,
        0.01073802]),
 'topic_2': array([0.89696629, 0.01335545, 0.01031684, ..., 0.01173731, 0.01183511,
        0.01073848]),
 'topic_3': array([0.011448  , 0.01335556, 0.01031708, ..., 0.01173817, 0.01183512,
        0.01073853]),
 'topic_4': array([0.0114489 , 0.01335492, 0.90714527, ..., 0.01173733, 0.01183695,
        0.90335547]),
 'topic_5': array([0.01144814, 0.01335488, 0.0103169 , ..., 0.01173728, 0.01183433,
        0.01073812]),
 'topic_6': array([0.01144837, 0.87980156, 0.01031722, ..., 0.01173746, 0.01183464,
        0.01073837]),
 'topic_7': array([0.01144809, 0.01335702, 0.01031717, ..., 0.01173743, 0.01183486,
        0.010738  ]),
 'topic_8': array([0.01144784, 0.01335487, 0.01031692, ..., 0.8943606 , 0.01183477,
        0.01073798]),
 'topic_9': array([0.01144852, 0.01335475, 0.0

In [57]:
ten_topics_df = pd.DataFrameme(ten_topics)
ten_topic_means = ten_topics_df.mean()
print(ten_topic_means)

topic_0    0.063823
topic_1    0.084326
topic_2    0.104312
topic_3    0.105723
topic_4    0.152280
topic_5    0.081898
topic_6    0.100417
topic_7    0.114937
topic_8    0.077470
topic_9    0.114814
dtype: float64


#### Topic mixture of words

In [58]:
ten_topic_word_mixture = pd.DataFrame(
    ten_lda_model.components_,
    columns = tf_idf_vectorizer.get_feature_names_out()
)

### Print topics

In [59]:
def print_topics(lda_model, vectorizer, top_words):
    # 1. TOPIC MIXTURE OF WORDS FOR EACH TOPIC
    topic_mixture = pd.DataFrame(
        lda_model.components_,
        columns = vectorizer.get_feature_names_out()
    )

    # 2. FINDING THE TOP WORDS FOR EACH TOPIC
    ## Number of topics
    n_components = topic_mixture.shape[0]

    ## Top words for each topic
    for topic in range(n_components):
        print("-"*10)
        print(f"For topic {topic}, here are the the top {top_words} words with weights:")

        topic_df = topic_mixture.iloc[topic]\
            .sort_values(ascending = False).head(top_words)

        print(round(topic_df,3))

### Finding topics for Bigrams only (10 topics)

In [ ]:
%%time

n_components = 10
bigram_lda_model = LatentDirichletAllocation(n_components=n_components, max_iter = 100)


bigram_lda_model.fit(bigram_combined_weighted_words)

#### Most relevant words for each topic

In [ ]:
print_topics(bigram_lda_model, bigram_tf_idf_vectorizer, 5)